src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -26,6 +27,36 @@ public sealed partial class NpgsqlReadBuffer
 
         internal Stream Underlying { private get; set; }
 
+        readonly Socket _underlyingSocket;
+
+        CancellationTokenSource _timeoutCts = new CancellationTokenSource();
+
+        /// <summary>
+        /// Underlying socket ReceiveTimeout
+        /// </summary>
+        internal TimeSpan Timeout
+        {
+            get => _currentSocketTimeout;
+            set
+            {
+                if (_currentSocketTimeout != value)
+                {
+                    if (_underlyingSocket == null)
+                        throw new InvalidOperationException("Unable to set timeout as underlying socket is null");
+
+                    _underlyingSocket.ReceiveTimeout = value > TimeSpan.Zero ? (int)value.TotalMilliseconds : -1;
+                    _currentSocketTimeout = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Contains the current value of the Socket's RecieveTimeout, used to determine whether
+        /// we need to change it when commands are received.
+        /// It's used only for sync IO (<see cref="Stream.Read(byte[], int, int)"/>)
```

Comment:
This comment doesn't seem right - isn't _currentSocketTimeout is used for async timeouts in Ensure (via the Timeout property?)

Reply:
Yes, it'not entirely correct. Rewrote it + renamed `_currentSocketTimeout` to `_currentTimeout`.